monai/handlers/iteration_metric.py

```diff
@@ -77,7 +77,7 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
         score = self.metric_fn(y_pred, y)
         if isinstance(score, (tuple, list)):
             score = score[0]
-        self._scores.append(score)
+        self._scores.append(score.clone().to(self._device))
```

Comment:
I think you can check the device of `score` and optionally pass it to `self._device` if mismatch. Otherwise cloning and sending can be expensive for each iteration. What do you think, @Nic-Ma ?

Reply:
Hi @vfdev-5 ,

I agree with you, so I think I don't need to update this line at all, just keep the original:
```py
self._scores.append(score)
```

Reply:
Well, in this case, `self._device` wont be used, right ?
If user sets `device="cpu"`, but function returns "cuda", internal storage will be "cuda" and not cpu as required by the user ?

Reply:
Hi @vfdev-5 ,

What's the expected behavior in ignite metrics? What is this `device` arg used for?
I refer to this code:
https://github.com/pytorch/ignite/blob/master/ignite/metrics/epoch_metric.py#L102

Thanks.

Reply:
Do you mean to add another arg to optionally pass `device` to `self._device`?

Thanks.

Reply:
Expected behaviour is to put metric internal tensors to `self._device`. On the line to you cite here, we clone y_pred and y as we do not manage the lifetime of those tensors (and user can potentially modify inplace them after...).
In your case, score is a tensor from user-defined function and its lifetime is managed by your class. I think you can just simply do `score.to(self._device)` (and use if to check if eventually need to put to `self._device`).

Reply:
Hi @vfdev-5 ,

I updated the PR according to your suggestion.
I think PyTorch will not copy the Tensor if the device of `score` is same as `self._device`:
https://pytorch.org/docs/stable/tensors.html#torch.Tensor.to

Thanks.

Reply:
Yes, I think so.